lib/pbio/sys/bluetooth.c

```diff
@@ -195,6 +210,28 @@ bool pbsys_bluetooth_tx_is_idle(void) {
     return !send_busy && lwrb_get_full(&stdout_ring_buf) == 0;
 }
 
+void pbsys_bluetooth_enabled_state_request_toggle(void) {
```

Comment:
Instead of all of these status checks and starting/stopping the bluetooth thread, I think we could just add some checks in the bluetooth thread at the appropriate times. Stopping the thread seems like an easy way to break things unintentionally.

```diff
diff --git a/lib/pbio/sys/bluetooth.c b/lib/pbio/sys/bluetooth.c
index 2eea566b..84268ee0 100644
--- a/lib/pbio/sys/bluetooth.c
+++ b/lib/pbio/sys/bluetooth.c
@@ -290,6 +290,9 @@ PROCESS_THREAD(pbsys_bluetooth_process, ev, data) {
     pbdrv_bluetooth_set_receive_handler(handle_receive);
 
     while (!pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN)) {
+        // TODO: may need to also check shutdown status here
+        PROCESS_WAIT_WHILE(pbsys_status_test(PBIO_PYBRICKS_STATUS_BLE_OFF));
+
         // make sure the Bluetooth chip is in reset long enough to actually reset
         etimer_set(&timer, 150);
         PROCESS_WAIT_EVENT_UNTIL(ev == PROCESS_EVENT_TIMER && etimer_expired(&timer));
@@ -305,7 +308,8 @@ PROCESS_THREAD(pbsys_bluetooth_process, ev, data) {
         PROCESS_WAIT_UNTIL(
             pbdrv_bluetooth_is_connected(PBDRV_BLUETOOTH_CONNECTION_LE)
             || pbsys_status_test(PBIO_PYBRICKS_STATUS_USER_PROGRAM_RUNNING)
-            || pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN));
+            || pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN)
+            || pbsys_status_test(PBIO_PYBRICKS_STATUS_BLE_OFF));
 
         if (!pbdrv_bluetooth_is_connected(PBDRV_BLUETOOTH_CONNECTION_LE)) {
             // if a connection wasn't made, we need to manually stop advertising
@@ -314,6 +318,10 @@ PROCESS_THREAD(pbsys_bluetooth_process, ev, data) {
 
         pbsys_status_clear(PBIO_PYBRICKS_STATUS_BLE_ADVERTISING);
 
+        if (pbsys_status_test(PBIO_PYBRICKS_STATUS_BLE_OFF)) {
+            goto ble_off;
+        }
+
         PT_INIT(&status_monitor_pt);
 
         while (pbdrv_bluetooth_is_connected(PBDRV_BLUETOOTH_CONNECTION_LE)
@@ -351,6 +359,7 @@ PROCESS_THREAD(pbsys_bluetooth_process, ev, data) {
             PROCESS_WAIT_EVENT();
         }
 
+ble_off:
         reset_all();
         PROCESS_WAIT_WHILE(pbsys_status_test(PBIO_PYBRICKS_STATUS_USER_PROGRAM_RUNNING));
 
```

Reply:
Thank you! I ended up doing something similar last night. We can even omit the `goto` and achieve the same by testing the BLE active status in that loop.